File: sdk/cpp/src/noise/symmetric_state.cpp
#include "noise/symmetric_state.hpp"

#include <algorithm>

namespace relayly::noise {

SymmetricState SymmetricState::Initialize(const std::string& protocol_name) {
  SymmetricState s;
  std::vector<std::uint8_t> name_bytes(protocol_name.begin(), protocol_name.end());
  if (name_bytes.size() <= kKeyLen) {
    std::fill(s.h_.begin(), s.h_.end(), std::uint8_t{0});
    std::copy(name_bytes.begin(), name_bytes.end(), s.h_.begin());
  } else {
    s.h_ = Hash(name_bytes);
  }
  s.ck_ = s.h_;
  return s;
}

void SymmetricState::MixKey(const std::vector<std::uint8_t>& input_key_material) {
  auto [new_ck, temp_k] = Hkdf2(ck_, input_key_material);
  ck_ = new_ck;
  cipher_state_.InitializeKey(temp_k);
}

void SymmetricState::MixHash(const std::vector<std::uint8_t>& data) {
  std::vector<std::uint8_t> combined(h_.begin(), h_.end());
  combined.insert(combined.end(), data.begin(), data.end());
  h_ = Hash(combined);
}

std::vector<std::uint8_t> SymmetricState::EncryptAndHash(const std::vector<std::uint8_t>& plaintext) {
  std::vector<std::uint8_t> ad(h_.begin(), h_.end());
  auto ciphertext = cipher_state_.EncryptWithAd(ad, plaintext);
  MixHash(ciphertext);
  return ciphertext;
}

std::vector<std::uint8_t> SymmetricState::DecryptAndHash(const std::vector<std::uint8_t>& ciphertext) {
  std::vector<std::uint8_t> ad(h_.begin(), h_.end());
  auto plaintext = cipher_state_.DecryptWithAd(ad, ciphertext);
  MixHash(ciphertext);
  return plaintext;
}

std::pair<CipherState, CipherState> SymmetricState::Split() {
  auto [temp_k1, temp_k2] = Hkdf2(ck_, {});
  CipherState c1;
  CipherState c2;
  c1.InitializeKey(temp_k1);
  c2.InitializeKey(temp_k2);
  return {c1, c2};
}

}  // namespace relayly::noise

File: sdk/ts/src/noise/symmetricState.ts
import { CipherState } from './cipherState.js';
import { HASHLEN, hash, noiseHKDF } from './primitives.js';

const PROTOCOL_NAME = new TextEncoder().encode('Noise_XX_25519_ChaChaPoly_BLAKE2s');

function concat(...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const p of parts) {
    out.set(p, offset);
    offset += p.length;
  }
  return out;
}

/** Noise's SymmetricState: (ck, h, cipherState), threaded through one handshake. */
export class SymmetricState {
  private ck: Uint8Array;
  private h: Uint8Array;
  private readonly cipherState = new CipherState();

  constructor() {
    // InitializeSymmetric(protocol_name): "Noise_XX_25519_ChaChaPoly_BLAKE2s" is 34
    // bytes, longer than HASHLEN (32), so h = HASH(protocol_name); ck = h.
    this.h = PROTOCOL_NAME.length <= HASHLEN
      ? concat(PROTOCOL_NAME, new Uint8Array(HASHLEN - PROTOCOL_NAME.length))
      : hash(PROTOCOL_NAME);
    this.ck = this.h.slice();
  }

  mixKey(inputKeyMaterial: Uint8Array): void {
    const [ck, tempK] = noiseHKDF(this.ck, inputKeyMaterial, 2);
    this.ck = ck;
    this.cipherState.initializeKey(tempK);
  }

  mixHash(data: Uint8Array): void {
    this.h = hash(concat(this.h, data));
  }

  encryptAndHash(plaintext: Uint8Array): Uint8Array {
    const ciphertext = this.cipherState.encryptWithAd(this.h, plaintext);
    this.mixHash(ciphertext);
    return ciphertext;
  }

  decryptAndHash(ciphertext: Uint8Array): Uint8Array {
    const plaintext = this.cipherState.decryptWithAd(this.h, ciphertext);
    this.mixHash(ciphertext);
    return plaintext;
  }

  /** Split(): derives the two transport cipher states once the handshake completes. */
  split(): [CipherState, CipherState] {
    const [k1, k2] = noiseHKDF(this.ck, new Uint8Array(0), 2);
    const c1 = new CipherState();
    c1.initializeKey(k1);
    const c2 = new CipherState();
    c2.initializeKey(k2);
    return [c1, c2];
  }

  getHandshakeHash(): Uint8Array {
    return this.h;
  }
}
